### ISSUE TITLE
Segmentation fault when looping with `itertools.groupby`

### ISSUE THREAD

**Issue by sauliusl** [daler/pybedtools#131]
Looping through a `BedTool` with `groupby` seems to generate a segfault.
For instance, to reproduce do this:

```python
import pybedtools
from itertools import groupby

x = pybedtools.BedTool([('chr1', 12, 13, 'N', 1000, '+'), 
                        ('chr1', 12, 13, 'N', 1000, '-'), 
                        ('chr1', 12, 13, 'N', 1000, '-'), 
                        ('chr1', 115, 116, 'N', 1000, '+')])

for key, group_ in groupby(x, key=lambda r: (r.chrom, r.start, r.end)):
    print key, map(lambda r: r.strand, group_)
```
This would generate the following output:
```
('chr1', 12, 13) ['+', '-', '-']
('chr1', 115, 116) ['+']
Segmentation fault (core dumped)
```
Looks like there are some issues when accessing the `Interval` objects within the `group_`.

For instance, accessing the start variable 
```python
for key, group_ in groupby(x, key=lambda r: (r.chrom, r.start, r.end)):
    print key, map(lambda r: r.start, group_)
```
would fail in a very similar fashion:
```
('chr1', 12, 13) [12, 12, 12]
('chr1', 115, 116) [115]
terminate called after throwing an instance of 'std::bad_cast'
  what():  std::bad_cast
Aborted (core dumped)
```
Or, just converting the group to list:
```python
for key, group_ in groupby(x, key=lambda r: (r.chrom, r.start, r.end)):    
    list(group_)
```
fails in a very similar fashion:
```
terminate called after throwing an instance of 'std::bad_cast'
  what():  std::bad_cast
Aborted (core dumped)
```

I would suspect some double destruction happening somewhere maybe?

**Comment by daler** [comment]
Thanks for reporting this. My guess is that `cbedtools.IntervalIterator` is doing something funky, like making a double iterator or something (along the lines of your double destruction idea).  I'll look into it.

Depending on what you need to do, you could try `pybedtools.BedTool.groupby` in the meantime.

**Comment by daler** [comment]
While I haven't had the time to debug in detail, it looks like you can use a generator function as a temporary workaround until this is fixed:

```python
import pybedtools
from itertools import groupby

x = pybedtools.BedTool([('chr1', 12, 13, 'N', 1000, '+'), 
                        ('chr1', 12, 13, 'N', 1000, '-'), 
                        ('chr1', 12, 13, 'N', 1000, '-'), 
                        ('chr1', 115, 116, 'N', 1000, '+')])
def gen():
    for i in x:
        yield i

for key, group_ in groupby(gen(), key=lambda r: (r.chrom, r.start, r.end)):
    print key, map(lambda r: r.strand, group_)
```

**Comment by csfoo** [comment]
I ran into this issue and I think that the issue is in the iterator protocol implemented by IntervalFile; this issue is now fixed since BedTool iterators are now derived from IntervalIterator. 

Specifically, python's iterator protocol requires "that once an iterator’s next() method raises StopIteration, it will continue to do so on subsequent calls". IntervalFile's iterator does not. 

This property is used by the [itertools.groupby code](https://docs.python.org/2/library/itertools.html#itertools.groupby) to determine when to quit. The first time IntervalFile throws StopIteration (in _grouper), it closes the intervalFile_ptr. However, groupby.next tries to call next on IntervalFile's iterator again, and then it will attempt to read from the now closed file and segfault on GetNextBed (_bedStream->good()).
